geopandas/plotting.py

```diff
@@ -582,10 +582,21 @@ def plot_dataframe(
         # set categorical to True for creating the legend
         categorical = True
         binedges = [values[~nan_idx].min()] + binning.bins.tolist()
-        categories = [
-            "{0:.2f} - {1:.2f}".format(binedges[i], binedges[i + 1])
-            for i in range(len(binedges) - 1)
-        ]
+        if legend_kwds is not None and 'labels' in legend_kwds:
+            if len(legend_kwds['labels']) != classification_kwds['k']:
+                raise AttributeError('Number of labels must match number of bins, '
+                                     'received {} labels for {} bins'.
+                                     format(len(legend_kwds['labels']),
+                                            classification_kwds['k']))
+            else:
+                categories = list(legend_kwds['labels'])
```

Comment:
Use `legend_kwds.pop('labels')` here, then you don't need `del legend_kwds['labels']` below.